File: isSymmetric.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function (root) {
  let isMirror = (t1, t2) => {
    if (t1 == null && t2 == null) return true;
    if (t1 == null || t2 == null) return false;
    return (
      t1.val == t2.val &&
      isMirror(t1.right, t2.left) &&
      isMirror(t1.left, t2.right)
    );
  };
  return isMirror(root, root);
};

var isSymmetric = function (root) {
  let stack = [root, root];
  while (stack.length) {
    let t1 = stack.pop();
    let t2 = stack.pop();
    if (t1 == null || t2 == null) return false;
    if (t1.val != t2.val) return false;
    if (t1 && t2) {
      stack.push(t1.left);
      stack.push(t2.right);
      stack.push(t1.right);
      stack.push(t2.left);
    }
  }
  return isMirror(root, root);
};

// public boolean isSymmetric(TreeNode root) {
//   Queue<TreeNode> q = new LinkedList<>();
//   q.add(root);
//   q.add(root);
//   while (!q.isEmpty()) {
//       TreeNode t1 = q.poll();
//       TreeNode t2 = q.poll();
//       if (t1 == null && t2 == null) continue;
//       if (t1 == null || t2 == null) return false;
//       if (t1.val != t2.val) return false;
//       q.add(t1.left);
//       q.add(t2.right);
//       q.add(t1.right);
//       q.add(t2.left);
//   }
//   return true;
// }

// class Solution:
//     def isSymmetric(self, root):
//         stack = []
//         if root: stack.append([root.left, root.right])

//         while(len(stack) > 0):
//             left, right = stack.pop()
            
//             if left and right:
//                 if left.val != right.val: return False
//                 stack.append([left.left, right.right])
//                 stack.append([right.left, left.right])
        
//             elif left or right: return False
        
//         return True

File: leetcode/isSymmetric.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        stack = []
        if root:
            stack.append([root.left, root.right])
        while(stack):
            left, right = stack.pop()

            if left and right:
                if left.val != right.val:
                    return False
                stack.append([left.left, right.right])
                stack.append([right.left, left.right])

            elif left or right:
                return False

        return True


def is_symmetric_dfs(root):
    def dfs(node1, node2):
        if node1 is None and node2 is None:
            return True
        if node1 is None or node2 is None:
            return False
        return node1.val == node2.val and dfs(
            node1.left, node2.right) and dfs(
            node1.right, node2.left)
    return dfs(root, root)
